lms/djangoapps/instructor/services.py

```diff
@@ -78,13 +77,61 @@ def delete_student_attempt(self, student_identifier, course_id, content_id, requ
                 err_msg = (
                     'Error occurred while attempting to reset student attempts for user '
-                    '{student_identifier} for content_id {content_id}.'.format(
-                        student_identifier=student_identifier,
-                        content_id=content_id
-                    )
+                    f'{student_identifier} for content_id {content_id}.'
                 )
                 log.error(err_msg)
 
+    def complete_student_attempt(self, user_identifier, content_id):
+        """
+        Submits all completable xblocks inside of the content_id block to the
+        Completion Service to mark them as complete. One use case of this function is
+        for special exams (timed/proctored) where regardless of submission status on
+        individual problems, we want to mark the entire exam as complete when the exam
+        is finished.
+
+        params:
+            user_identifier (String): username or email of a user
+            content_id (String): the block key for a piece of content
+        """
+        err_msg_prefix = (
+            'Error occurred while attempting to complete student attempt for user '
+            f'{user_identifier} for content_id {content_id}. '
+        )
+        err_msg = None
+        try:
+            user = get_user_by_username_or_email(user_identifier)
+            block_key = UsageKey.from_string(content_id)
+            root_block = modulestore().get_item(block_key)
+        except ObjectDoesNotExist:
+            err_msg = err_msg_prefix + 'User does not exist!'
+        except InvalidKeyError:
+            err_msg = err_msg_prefix + 'Invalid content_id!'
+        except ItemNotFoundError:
+            err_msg = err_msg_prefix + 'Block not found in the modulestore!'
+        if err_msg:
+            log.error(err_msg)
+            return
+
+        def _submit_completions(block, user):
+            """
+            Recursively submits the children for completion to the Completion Service
+            """
+            mode = XBlockCompletionMode.get_mode(block)
+            if mode == XBlockCompletionMode.COMPLETABLE:
+                block.runtime.publish(block, 'completion', {'completion': 1.0, 'user_id': user.id})
+            elif mode == XBlockCompletionMode.AGGREGATOR:
+                # I know this looks weird, but at the time of writing at least, there isn't a good
+                # single way to get the children assigned for a partcular user. Some blocks define the
+                # child descriptors method, but others don't and with blocks like Randomized Content
+                # (Library Content), the get_children method returns all children and not just assigned
+                # children. So this is our way around situations like that.
+                block_children = ((hasattr(block, 'get_child_descriptors') and block.get_child_descriptors())
```

Comment:
do you think this code is useful enough to go in some utility class/file where it could be reused by other code?

Reply:
I am honestly not sure. I don't know how often this situation is run into and it feels like both of the functions I'm calling out to here are already trying to solve the same problem. I'm hesitant to end up adding a third method and confuse people even more and I can't easily deprecate these two functions in favor of the "function to rule them all" because of how widely used these functions are across our code base. So basically, /shrug